Add initial support for HTC One (M8)

amessier opened the pull request:
Add initial support for the HTC One (M8).

UART support is TBD, for now USB Networking can be used to access the Linux console.

Initial hardware supported:
 - eMMC
 - Power button
 - USB
 - Vibrator
 - Volume buttons (GPIO)
 - Wi-Fi

amessier committed:
ARM: dts: qcom: Add initial support for HTC One (M8)

Add initial device tree for the HTC One (M8) smartphone.

Initial support includes:
 - eMMC
 - Power button
 - USB
 - Vibrator
 - Volume buttons (GPIO)
 - Wi-Fi

Signed-off-by: Alexandre Messier <alex@me.ssier.org>

amessier committed:
dt-bindings: arm: qcom: Add HTC One (M8)

Add a compatible for the HTC One (M8), which is based on the msm8974pro.

Signed-off-by: Alexandre Messier <alex@me.ssier.org>

z3ntu commented on arch/arm/boot/dts/qcom/qcom-msm8974-htc-m8.dts:
```diff
@@ -0,0 +1,332 @@
+	};
+
+	chosen {
+	};
```
Remove this empty node if not needed

amessier replied:
I had to put an empty `chosen` node, for lk2nd. It does not create it if it absent.

Usually there is a `stdout-path`, but since I don't know which UART is the debug UART, there is nothing else to put here. Should I add a comment in the empty node regarding this?

z3ntu replied:
Does something break if there's no `chosen{};`?

amessier replied:
Yes if there is no `chosen` node, lk2nd will fail to edit the device tree before starting the kernel (due to [this code](https://github.com/msm8916-mainline/lk2nd/blob/b02d7582ef456ba1e0297d8a2beb3e098d1d8ce9/platform/msm_shared/dev_tree.c#L1406-L1411))

z3ntu replied:
Got it. I've fixed this in SoC dtsi now (and sent upstream, don't see why this shouldn't be accepted) https://github.com/msm8974-mainline/linux/commit/44c10594ebb8f0597b1cc5182d03b3acf2c92c7a so if you rebase you can drop it then :)

amessier replied:
OK rebased, tested and it works.

z3ntu commented on arch/arm/boot/dts/qcom/qcom-msm8974-htc-m8.dts:
```diff
@@ -0,0 +1,332 @@
+
+&pronto {
+	/* Use v1 to get correct vddmx-supply range. */
+	compatible = "qcom,pronto-v1-pil", "qcom,pronto";
```
Can you link to downstream devicetree why this is necessary? Does your device maybe come with a different WiFi/Bluetooth chip than WCN3680?

amessier replied:
Sure, please have a look [here](https://github.com/LineageOS/android_kernel_htc_msm8974/blob/lineage-18.1/drivers/net/wireless/wcnss/wcnss_vreg.c#L115).

I did not really investigate why the mismatch. When I saw that `pronto-v1-pil` matched the vendor kernel, I went with it.

According to the iFixit teardown, it is indeed a WCN3680, maybe an older/different revision?

z3ntu replied:
Does not overriding the compatible also work? I checked a bit and other 8974 devices seem to look the same as your M8

amessier replied:
I tried to keep the default compatible, and wifi does work. But I am seeing an error message in the logs: `s1: unsupportable voltage range: 1287500-1050000uV` which is probably why I investigated this and found that the `pronto-v1-pil` had a correct voltage range, same range as the vendor kernel.

Since it works with the compatible from the dtsi, I removed the override here.

z3ntu replied:
> s1: unsupportable voltage range: 1287500-1050000uV

Oh that's where this error is from :sweat_smile: I've seen this since forever on FP2 also but it's not a fatal error :) 

I'll make an issue for fixing this for all, since it should be applicable to all msm8974 devices with pronto. https://github.com/msm8974-mainline/linux/issues/15

z3ntu commented:
I'm assuming you checked all the regulator voltages against downstream? Generally looks good, just added some small comments.

I'm also encouraging you to send this dts upstream quite soon, I don't want this dts to get stuck in this repo only :)

amessier commented:
> I'm assuming you checked all the regulator voltages against downstream? Generally looks good, just added some small comments.

Yes indeed, I matched the regulator voltages against the vendor device trees.
 
> I'm also encouraging you to send this dts upstream quite soon, I don't want this dts to get stuck in this repo only :)

That was part of my next steps :) I tried this series on top of 6.8-rc but I did not a valid boot (no ssh access). 6.7 did work though, so maybe it's something small. I will re-verify with 6.9-rc when it's released.

z3ntu commented:
You can rebase on top of `qcom-msm8974-6.8.y` (if that already exists by the time you read this) or `qcom-msm8974-6.8-wip`, that branch seems to work fine on my devices.

amessier commented:
> You can rebase on top of `qcom-msm8974-6.8.y` (if that already exists by the time you read this) or `qcom-msm8974-6.8-wip`, that branch seems to work fine on my devices.

I rebased on top of `qcom-msm8974-6.8.y`. I had the same issue as when I tried mainline 6.8, turns out it was the `smbb` node not being enabled, and the USB was probe-deferred indefinitely thus no USB networking. But with that fixed, it works like on 6.7.

I also updated the device tree a bit:
- reordered the `pinctrl` lines to put the `pinctrl-names` last
- updated the gpio debounce value [as per vendor kernel value](https://github.com/LineageOS/android_kernel_htc_msm8974/blob/21c9ae566de3a9f67683040bfb0b76810c92940b/drivers/input/misc/gpio_event.c#L162-L166)
- fixed `make CHECK_DTBS=y W=1` warning on the `gpio_keys_default` node

z3ntu commented:
Thanks, pushed a tiny fixup on top.

But one major thing is still pending, I'm thinking M8 is actually based on msm8974pro, can you run this command on mainline and post the output? `tail /sys/devices/soc0/*`

amessier commented:
> But one major thing is still pending, I'm thinking M8 is actually based on msm8974pro, can you run this command on mainline and post the output? `tail /sys/devices/soc0/*`

You are correct, it's a `msm8974pro` variant `AB`.

Here is the output as requested:
```
htc-m8:~$ tail /sys/devices/soc0/*
==> /sys/devices/soc0/family <==
Snapdragon

==> /sys/devices/soc0/machine <==
MSM8974PRO-AB

==> /sys/devices/soc0/power <==
tail: read error: Is a directory

==> /sys/devices/soc0/revision <==
1.1

==> /sys/devices/soc0/serial_number <==
xxxxxxxxxx

==> /sys/devices/soc0/soc_id <==
218

==> /sys/devices/soc0/subsystem <==
tail: read error: Is a directory

==> /sys/devices/soc0/uevent <==
htc-m8:~$
```

Indeed, I originally thought of using `qcom-msm8974pro.dtsi` (due to the `pro` in the name matching the model), but it looks like that file is made for the `MSM8974PRO-AC` only, as it will enable the `AC` specific [faster  `sdhc1` clocks](https://github.com/torvalds/linux/blob/23956900041d968f9ad0f30db6dede4daccd7aa9/drivers/clk/qcom/gcc-msm8974.c#L871-L882).

In the [vendor kernel](https://github.com/LineageOS/android_kernel_htc_msm8974/blob/21c9ae566de3a9f67683040bfb0b76810c92940b/arch/arm/mach-msm/clock-8974.c#L1553-L1584)  those clocks are not used and even `#if 0`'d out.

So I don't think we can use `qcom-msm8974pro.dtsi` for the `AB` variant, which is why I used `qcom-msm8974.dtsi` instead (and matched the model name in the `dts` name).

z3ntu commented:
Nope, `qcom-msm8974pro.dtsi` is meant for Snapdragon 801 a.k.a MSM8974PRO.

MSM8974PRO-AC a.k.a MSM8974AB-AC (or basically whatever identifier was chosen that includes -AC) is another thing again (https://github.com/torvalds/linux/blob/master/arch/arm/boot/dts/qcom/qcom-msm8974pro-oneplus-bacon.dts#L190). But true, the clock driver seems to handle PRO-AA/PRO-AB and PRO-AC the same way which is probably not what's correct(?) https://github.com/torvalds/linux/commit/c685841ee127e8023a33be68fd8af6fe192a3665

But still, please rename your dtb to msm8974pro-htc-* and use msm8974pro.dtsi include. Also add msm8974pro compatible to the compatible list of the board, see other boards for example. Because it's definitely a PRO chip. I honestly don't think you'll see any problems because of this since we're using msm8974pro.dtsi for a while for also PRO-AA and PRO-AB devices.

nghfp9wa7bzq commented:
((Not sure, why pronto is needed at first glance...))  
Can you invite amessier to the chat please?  
Thanks.

z3ntu commented:
https://matrix.to/#/#msm8974-mainline:postmarketos.org - if joining through that doesn't work (let me know if it doesn't), you can try https://matrix.to/#/#msm8974-mainline:z3ntu.xyz which hopefully works.

amessier commented:
Sorry for the delay, I agree that the M8 is a 8974pro, the correct fix would maybe to introduce a new dtsi for the AB (like `qcom-msm8974pr-ab.dtsi`) which would use the `qcom,gcc-msm8974` compatible (or could also create `qcom,gcc-msm8974pro-ab` since a `-ac` compatible already exists).

In any case, I tried with the current situation, and the phone still boots and works, but you get a kernel warning splat:
```
[    0.787516] ------------[ cut here ]------------
[    0.787546] WARNING: CPU: 0 PID: 71 at drivers/clk/qcom/clk-rcg2.c:133 update_config+0xe4/0xf0
[    0.787590] sdcc1_apps_clk_src: rcg didn't update its configuration.
[    0.787598] Modules linked in:
[    0.787639] CPU: 0 PID: 71 Comm: kworker/u9:3 Not tainted 6.8.0-postmarketos-qcom-msm8974 #5
[    0.787669] Hardware name: Generic DT based system
[    0.787690] Workqueue: events_unbound async_run_entry_fn
[    0.787738] [<c01108f0>] (unwind_backtrace) from [<c010bc88>] (show_stack+0x10/0x14)
[    0.787784] [<c010bc88>] (show_stack) from [<c0c5faa4>] (dump_stack_lvl+0x40/0x4c)
[    0.787827] [<c0c5faa4>] (dump_stack_lvl) from [<c0121e60>] (__warn+0x78/0x154)
[    0.787873] [<c0121e60>] (__warn) from [<c0122118>] (warn_slowpath_fmt+0x1dc/0x1e4)
[    0.787916] [<c0122118>] (warn_slowpath_fmt) from [<c06a8b14>] (update_config+0xe4/0xf0)
[    0.787960] [<c06a8b14>] (update_config) from [<c06a8fc4>] (clk_rcg2_configure+0xa8/0xb0)
[    0.787999] [<c06a8fc4>] (clk_rcg2_configure) from [<c069b84c>] (clk_change_rate+0x98/0x548)
[    0.788040] [<c069b84c>] (clk_change_rate) from [<c069c1e0>] (clk_core_set_rate_nolock+0x1d8/0x3c0)
[    0.788078] [<c069c1e0>] (clk_core_set_rate_nolock) from [<c069c3f8>] (clk_set_rate+0x30/0x154)
[    0.788116] [<c069c3f8>] (clk_set_rate) from [<c09539b8>] (_opp_config_clk_single+0x24/0x84)
[    0.788165] [<c09539b8>] (_opp_config_clk_single) from [<c0956c6c>] (dev_pm_opp_set_rate+0x1c0/0x228)
[    0.788210] [<c0956c6c>] (dev_pm_opp_set_rate) from [<c098b154>] (sdhci_msm_probe+0x324/0xabc)
[    0.788258] [<c098b154>] (sdhci_msm_probe) from [<c07c1ab8>] (platform_probe+0x5c/0xb0)
[    0.788307] [<c07c1ab8>] (platform_probe) from [<c07beda8>] (really_probe+0xe0/0x3f8)
[    0.788346] [<c07beda8>] (really_probe) from [<c07bf15c>] (__driver_probe_device+0x9c/0x1f0)
[    0.788382] [<c07bf15c>] (__driver_probe_device) from [<c07bf2e0>] (driver_probe_device+0x30/0xc0)
[    0.788419] [<c07bf2e0>] (driver_probe_device) from [<c07bf418>] (__device_attach_driver+0xa8/0x120)
[    0.788455] [<c07bf418>] (__device_attach_driver) from [<c07bcdb4>] (bus_for_each_drv+0x84/0xd8)
[    0.788503] [<c07bcdb4>] (bus_for_each_drv) from [<c07be738>] (__device_attach_async_helper+0xac/0x100)
[    0.788546] [<c07be738>] (__device_attach_async_helper) from [<c014c8f0>] (async_run_entry_fn+0x40/0x158)
[    0.788590] [<c014c8f0>] (async_run_entry_fn) from [<c0141018>] (process_one_work+0x154/0x3c0)
[    0.788632] [<c0141018>] (process_one_work) from [<c0141500>] (worker_thread+0x27c/0x4ac)
[    0.788668] [<c0141500>] (worker_thread) from [<c0148a98>] (kthread+0xfc/0x12c)
[    0.788711] [<c0148a98>] (kthread) from [<c010014c>] (ret_from_fork+0x14/0x28)
[    0.788751] Exception stack(0xf0bf5fb0 to 0xf0bf5ff8)
[    0.788775] 5fa0:                                     00000000 00000000 00000000 00000000
[    0.788804] 5fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    0.788830] 5fe0: 00000000 00000000 00000000 00000000 00000013 00000000
[    0.788853] ---[ end trace 0000000000000000 ]---
```

In any case, if there isn't anything else, this should be good to go.

For the Matrix chatroom, I don't currently have an account, when I get one I'll join, thanks!

z3ntu commented:
This warning is (unfortunately) normal, so no worries about that. Shouldn't cause any functional issues I believe.

Please also send your patch upstream, if you need any help with that please let me know!

z3ntu reviewed (approved):
LGTM, thanks!

z3ntu commented:
Hi @amessier , any progress on sending the patches upstream?

amessier commented:
> Hi @amessier , any progress on sending the patches upstream?

@z3ntu sorry I don't have any progress, except I tested today with qcom for-next like you mentioned on Matrix. No issue, so I hope to send to the ML by next weekend. Thanks!

z3ntu commented:
Great to hear!
